### ISSUE TITLE
Feature: Eager loading in ActiveRecord::refresh()

### ISSUE THREAD

**Issue by chriscpty** [yiisoft/yii2#20277]
As of current, `ActiveRecord::refresh()` and `BaseActiveRecord::refresh()` only refresh the record itself and lose all relations. This is the expected behaviour, but if you then need to iterate over nested relations afterwards, lazy loading them leads to an excessive amount of DB queries.

My proposal would thus be to add an optional parameter `with` to `refresh`:

`public function refresh ($with = [])`

The implementation in `ActiveRecord` uses the corresponding model's `ActiveQuery` anyway, so it might as well call `->with(...$with)` on it in the process.

| Q                | A
| ---------------- | ---
| Yii version      | dev-master
| PHP version      | 8.1
| Operating system | Debian 12

I'd be happy to implement this, but want to be sure this makes sense and I haven't missed anything relevant before :)

**Comment by samdark** [comment]
I find it interesting. @yiisoft/yii2, @yiisoft/reviewers thoughts?

**Comment by bizley** [comment]
This would be great, but if we want to add this in Yii 2 we need to do this with method arguments discovery, like Symfony does sometimes (for BC).

**Comment by samdark** [comment]
What's "method arguments discovery"?

**Comment by chriscpty** [comment]
I don't think I see why that'd be necessary instead of just having the `$with` argument have a default value (which as far as I can tell is supported for all PHP versions yii supports)? Existing implementations that don't use this argument still work then, and while overrides may show a warning, they won't break either.

The only case of breaking I could see (that would happen afterwards though) is if a user overrides `refresh` and adds differently named arguments, then tries to pass `with` as a named argument - but I don't think that's a BC break.

**Comment by bizley** [comment]
Adding argument to refresh(), even optional one, will break user's apps that extend this method.

**Comment by schmunk42** [comment]
Should go into 2.2, if implemented.

**Comment by samdark** [comment]
2.2 may break compatibility.

**Comment by lubosdz** [comment]
Perhaps simpler variant without BC break, but somewhat less flexible .. ?

`public function refresh ($relations = true)`

true = default, no BC break
false = don't clear relations

**Comment by chriscpty** [comment]
I don't think that solves the BC break? The compatibility issue is that overrides of this function that don't have the additional optional parameter break.

**Comment by mtangoo** [comment]
So I see two options: 1. Add it in 2.0.x the way @bizley suggested or only implement in 2.2.x where such BC breaking is allowable. @chriscpty are you open to do it Bizley way?

**Comment by chriscpty** [comment]
I'd say I try to do both, one without the BC break and one with, and do separate pull requests? Then one can be merged into 2.0.x (so we can use it already) and one can be merged into 2.2.x (for a cleaner solution)?

**Comment by rob006** [comment]
Do we really need `$with` param? The fact that `refresh()` do not respect settings from `with()` looks more like a bug to me and IMO it should respect it by default.

**Comment by chriscpty** [comment]
I don't think I understand? Do you mean that `refresh` should eagerly load all relations that were loaded before calling it (for example through `with()` on the query the record was loaded with)? in which case, that's worth a thought too, but I'd personally rather have the control to decide which relations I want to load when refreshing

**Comment by rob006** [comment]
Yes, IMO these 2 samples should work the same:
 
```php
$model = MyModel::find()->with($withSettings)->where(['id' => $id])->one();
$model->name = 'test';
$model->save();
$model->refresh();
```

```php
$model = MyModel::find()->with($withSettings)->where(['id' => $id])->one();
$model->name = 'test';
$model->save();
$model = MyModel::find()->with($withSettings)->where(['id' => $id])->one();
```

**Comment by chriscpty** [comment]
That could also be an alternative, but i'd rather have control over which relations are loaded in the `refresh` independently of the relations loaded before.

Otherwise, you're forced to always reload relations even if you don't need them anymore afterwards

**Comment by rob006** [comment]
We could add new method `refreshWith($relations)`. Then you could overwrite `with()` settings using `$refresh` param (and pass empty array if you don't want to load relations).

**Comment by mtangoo** [comment]
> We could add new method `refreshWith($relations)`. Then you could overwrite `with()` settings using `$refresh` param (and pass empty array if you don't want to load relations).

That is another good option with total freedom. But if current method can be extended without affecting existing code (extending the classes and what not) then I would go for that first. 

What you guys think? @samdark @lubosdz @schmunk42 @bizley

**Comment by lubosdz** [comment]
Adding new method seems cleaner, as it is usual way of extending core functionality with @since docvar. 
Args discovery might not be quite cheap & simple.

**Comment by MarkoNV** [comment]
Do we really need another method just to be able to do
`
$model->refreshWith($relations);
`
instead of
`
$model->refresh();
$model->loadRelations($relations);
`

**Comment by chriscpty** [comment]
I wasn't aware of `loadRelations` when I first filled this issue, but yeah, that entirely fulfills this use case

**Comment by rob006** [comment]
`refreshWith()` would make sens if `refresh()` respected relations settings from initial query. I still think this is a bug that `refresh()` ignores relations config (https://github.com/yiisoft/yii2/issues/20277#issuecomment-2523237174).

**Comment by MarkoNV** [comment]
> `refreshWith()` would make sens if `refresh()` respected relations settings from initial query. I still think this is a bug that `refresh()` ignores relations config ([#20277 (comment)](https://github.com/yiisoft/yii2/issues/20277#issuecomment-2523237174)).

I think this is unrelated to this feature request.

BTW, regarding your idea, $relations is property of the query, not the model so I don't think it viable. Also, $relations can have callbacks, multilevel loading etc.

What would happened in case of:
```
$model = Post::find()->with('images.thumbs')->where(['id' => $id])->one();
$model->name = 'test';
$model->save();
foreach ($model->images as $image) {
    $image->refresh();
}

```
Should $image->thumbs be reloaded too? 

How about 
```
$model = Post::find()->with('images.thumbs')->where(['id' => $id])->one();
$model->loadRelations(['revisions']);
$model->refresh();
```
What should be reloaded? Imges and thumbs or revisions too?

**Comment by samdark** [comment]
Alright. If it's not related to this issue, I'm closing it. Feel free to open new one if necessary.
